modules/message-lane/src/outbound_lane.rs

```diff
@@ -0,0 +1,269 @@
+	/// Update lane data in the storage.
+	fn set_data(&mut self, data: OutboundLaneData);
+	/// Returns saved outbound message payload.
+	#[cfg(test)]
+	fn message(&self, nonce: &MessageNonce) -> Option<Self::Payload>;
+	/// Save outbound message in the storage.
+	fn save_message(&mut self, nonce: MessageNonce, payload: Self::Payload);
+	/// Remove outbound message from the storage.
+	fn remove_message(&mut self, nonce: &MessageNonce);
+}
+
+/// Outbound messages lane.
+pub struct OutboundLane<Storage> {
+	storage: Storage,
+}
+
+impl<Storage: OutboundLaneStorage> OutboundLane<Storage> {
+	/// Create new inbound lane backed by given storage.
+	pub fn new(storage: Storage) -> Self {
+		OutboundLane { storage }
+	}
+
+	/// Send message over lane.
+	///
+	/// Returns new message nonce.
+	pub fn send_message(&mut self, payload: Storage::Payload) -> MessageNonce {
+		let mut data = self.storage.data();
+		let nonce = data.latest_generated_nonce + 1;
+		data.latest_generated_nonce = nonce;
+
+		self.storage.save_message(nonce, payload);
+		self.storage.set_data(data);
+
+		nonce
+	}
+
+	/// Confirm message receival.
+	///
+	/// Returns `None` if confirmation is wrong/duplicate.
+	/// Returns `Some` with inclusive ranges of message nonces that have been received.
+	pub fn confirm_receival(&mut self, latest_received_nonce: MessageNonce) -> Option<(MessageNonce, MessageNonce)> {
+		let mut data = self.storage.data();
+		if latest_received_nonce <= data.latest_received_nonce || latest_received_nonce > data.latest_generated_nonce {
+			return None;
+		}
+
+		let prev_latest_received_nonce = data.latest_received_nonce;
+		data.latest_received_nonce = latest_received_nonce;
+		self.storage.set_data(data);
+
+		Some((prev_latest_received_nonce + 1, latest_received_nonce))
+	}
+
+	/// Confirm message processing.
+	///
+	/// Returns `None` if confirmation is wrong/duplicate.
+	/// Returns `Some` with inclusive ranges of message nonces that have been processed.
+	pub fn confirm_processing(&mut self, latest_processed_nonce: MessageNonce) -> Option<(MessageNonce, MessageNonce)> {
+		let mut data = self.storage.data();
+		// wait for recieval confirmation first
+		if latest_processed_nonce <= data.latest_processed_nonce || latest_processed_nonce > data.latest_received_nonce
+		{
+			return None;
+		}
+
+		let prev_latest_processed_nonce = data.latest_processed_nonce;
+		data.latest_processed_nonce = latest_processed_nonce;
+		self.storage.set_data(data);
+
+		Some((prev_latest_processed_nonce + 1, latest_processed_nonce))
+	}
+
+	/// Prune at most `max_messages_to_prune` already received messages.
+	pub fn prune_messages(&mut self, max_messages_to_prune: MessageNonce) -> MessageNonce {
+		let mut pruned_messages = 0;
+		let mut anything_changed = false;
+		let mut data = self.storage.data();
+		while pruned_messages < max_messages_to_prune && data.oldest_unpruned_nonce <= data.latest_received_nonce {
+			self.storage.remove_message(&data.oldest_unpruned_nonce);
```

Comment:
I wonder if there's a way to use a double map or something to more efficiently remove these messages

Reply:
Even if we would have the most natural API for this structure (I mean queue), I doubt the number of writes would be the less than `N`. Unless we'll store it as `Vec<_>` as Tomek has suggested.

Reply:
I didn't realize you and Tomek had already had a similar conversation. I'm okay with whatever outcome comes from that

Reply:
As suggested in some other comment, I'm more inclined to prune in `on_initialize` and have the costs be amortized.